File: technical_architecture/scripts/train_physics_rf.py
#!/usr/bin/env python3
"""
Train Random Forest on Physics Features (46D) for Hybrid Expert Ensemble
========================================================================

Extracts 46D physics features (Layer 1) from the 112D cache and trains
a Random Forest classifier for the Hybrid Expert ensemble.

Usage:
    python3 scripts/train_physics_rf.py
"""

import json
import numpy as np
import struct
from pathlib import Path
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score, classification_report
import joblib
import time

# Feature dimensions
FEATURE_DIM = 112
PHYSICS_DIM = 46  # Layer 1: indices 0-45

def load_bincode_features(filepath):
    """Load features stored in Rust bincode format (Vec<f32>)"""
    with open(filepath, 'rb') as f:
        # Read length as varint (bincode uses varint encoding)
        length = 0
        shift = 0
        while True:
            byte = struct.unpack('B', f.read(1))[0]
            length |= (byte & 0x7F) << shift
            shift += 7
            if byte & 0x80 == 0:
                break
        # Read features
        data = f.read(length * 4)  # 4 bytes per f32
        features = np.frombuffer(data, dtype=np.float32)
        return features.copy()  # Copy to make writable

def main():
    print("╔═══════════════════════════════════════════════════════════════════╗")
    print("║  Random Forest Training - Physics Features (46D)                 ║")
    print("║  For Hybrid Expert Ensemble                                       ║")
    print("╚═══════════════════════════════════════════════════════════════════╝")
    print()

    start_time = time.time()

    # Load manifest
    print("Loading manifest...")
    with open("beans_zero_full_manifest.json") as f:
        manifest = json.load(f)

    samples = manifest["samples"]
    print(f"  Total samples: {len(samples)}")

    # Load cache manifest
    with open("beans_feature_cache_112d/cache_manifest.json") as f:
        cache_manifest = json.load(f)

    print(f"  Cached features: {len(cache_manifest['entries'])}")

    # Load all features and labels
    print("\nLoading features from cache...")
    all_features = []
    all_labels = []

    cache_dir = Path("beans_feature_cache_112d")

    for sample in samples:
        audio_file = sample["audio_file"]
        label = sample["labels"]["output"] if sample["labels"]["output"] != "None" else f"task_{sample['labels']['task']}"

        cache_file = cache_manifest["entries"].get(audio_file)
        if cache_file:
            full_path = cache_dir / cache_file
            if full_path.exists():
                try:
                    features = load_bincode_features(full_path)
                    if len(features) == FEATURE_DIM:
                        # Extract only physics features (first 46D)
                        physics = features[:PHYSICS_DIM]
                        all_features.append(physics)
                        all_labels.append(label)
                except Exception as e:
                    pass  # Skip problematic files

    print(f"  Loaded {len(all_features)} samples")

    if len(all_features) == 0:
        raise ValueError("No features loaded!")

    # Convert to numpy arrays
    X = np.array(all_features)
    y = np.array(all_labels)

    print(f"  Feature shape: {X.shape}")
    print(f"  Unique labels: {len(np.unique(y))}")

    # Split into train/test (90/10)
    n_samples = len(X)
    n_train = int(n_samples * 0.9)

    # Shuffle
    indices = np.random.permutation(n_samples)
    train_idx = indices[:n_train]
    test_idx = indices[n_train:]

    X_train, y_train = X[train_idx], y[train_idx]
    X_test, y_test = X[test_idx], y[test_idx]

    print(f"\nSplit: {len(X_train)} train, {len(X_test)} test")

    # Standardize features
    print("\nStandardizing features...")
    scaler = StandardScaler()
    X_train_scaled = scaler.fit_transform(X_train)
    X_test_scaled = scaler.transform(X_test)

    # Train Random Forest with balanced class weights
    print("\nTraining Random Forest...")
    print("  n_estimators: 200")
    print("  max_depth: 30")
    print("  min_samples_split: 5")
    print("  class_weight: balanced")

    rf = RandomForestClassifier(
        n_estimators=200,
        max_depth=30,
        min_samples_split=5,
        class_weight='balanced',
        n_jobs=-1,
        random_state=42,
        verbose=1
    )

    rf.fit(X_train_scaled, y_train)

    # Evaluate
    print("\n╔═══════════════════════════════════════════════════════════════════╗")
    print("║  Evaluation Results                                               ║")
    print("╚═══════════════════════════════════════════════════════════════════╝")
    print()

    y_pred = rf.predict(X_test_scaled)
    accuracy = accuracy_score(y_test, y_pred) * 100

    print(f"Test Accuracy: {accuracy:.2f}%")
    print(f"Correct: {np.sum(y_pred == y_test)} / {len(y_test)}")

    # Top 20 classes
    print("\nTop 20 Classes by Sample Count:")
    print(f"{'Class':<50} {'Total':>8} {'Correct':>8} {'Accuracy':>8}")
    print("-" * 76)

    unique_labels, counts = np.unique(y_test, return_counts=True)
    label_stats = list(zip(unique_labels, counts))
    label_stats.sort(key=lambda x: -x[1])

    for label, total in label_stats[:20]:
        mask = y_test == label
        correct = np.sum(y_pred[mask] == label)
        acc = correct / total * 100 if total > 0 else 0
        print(f"{label:<50} {total:>8} {correct:>8} {acc:>7.1f}%")

    # Save model
    model_path = "physics_rf_model.joblib"
    print(f"\nSaving model to: {model_path}")
    joblib.dump({
        'model': rf,
        'scaler': scaler,
        'feature_dim': PHYSICS_DIM,
        'accuracy': accuracy,
    }, model_path)

    # Summary
    elapsed = time.time() - start_time
    print("\n╔═══════════════════════════════════════════════════════════════════╗")
    print("║  Summary                                                          ║")
    print("╠═══════════════════════════════════════════════════════════════════╣")
    print(f"║  Architecture:       Random Forest (Physics 46D)                 ║")
    print(f"║  Test Accuracy:      {accuracy:>8.2f}%                                   ║")
    print(f"║  Total Time:         {elapsed:>8.1f}s                                    ║")
    print("╚═══════════════════════════════════════════════════════════════════╝")

    print("\nHybrid Expert Ensemble Summary:")
    print(f"  - Texture NN (66D):  59.88%")
    print(f"  - Physics RF (46D):  {accuracy:.2f}%")
    print("  - Combined ensemble: (run ensemble evaluation)")

    return accuracy

if __name__ == "__main__":
    main()

File: technical_architecture/src/bin/train_physics_rf.rs
//! Train Random Forest on Physics Features (46D) for Hybrid Expert Ensemble
//! ========================================================================
//!
//! Trains a Random Forest classifier on the 46D physics features (Layer 1)
//! to be combined with the Texture NN for the Hybrid Expert ensemble.
//!
//! Usage:
//!   cargo run --release --bin train_physics_rf --features ml-classical

use anyhow::Result;
use linfa::prelude::*;
use linfa_trees::DecisionTree;
use linfa_preprocessing::linear_scaling::LinearScaler;
use ndarray::{Array1, Array2};
use serde::Deserialize;
use std::collections::HashMap;
use std::fs;
use std::io::BufReader;
use std::path::Path;
use std::time::Instant;

use technical_architecture::taxonomic_router::{
    FEATURE_DIM, PHYSICS_DIM, slice_physics,
};

// =============================================================================
// Data Structures
// =============================================================================

#[derive(Debug, Deserialize)]
struct BeansManifest {
    #[allow(dead_code)]
    dataset: String,
    #[allow(dead_code)]
    n_samples: usize,
    samples: Vec<BeansSample>,
}

#[derive(Debug, Deserialize, Clone)]
struct BeansSample {
    audio_file: String,
    #[allow(dead_code)]
    n_samples: u32,
    labels: BeansLabels,
}

#[derive(Debug, Deserialize, Clone)]
struct BeansLabels {
    output: String,
    task: String,
}

#[derive(Debug, Deserialize)]
struct CacheManifest {
    entries: HashMap<String, String>,
    #[allow(dead_code)]
    feature_count: usize,
}

// =============================================================================
// Main
// =============================================================================

fn main() -> Result<()> {
    println!("╔═══════════════════════════════════════════════════════════════════╗");
    println!("║  Random Forest Training - Physics Features (46D)                 ║");
    println!("║  For Hybrid Expert Ensemble                                       ║");
    println!("╚═══════════════════════════════════════════════════════════════════╝");
    println!();

    let start = Instant::now();

    // Load manifest
    let manifest_path = "beans_zero_full_manifest.json";
    println!("Loading manifest from: {}", manifest_path);
    let manifest_data = fs::read_to_string(manifest_path)?;
    let manifest: BeansManifest = serde_json::from_str(&manifest_data)?;
    println!("  Total samples: {}", manifest.samples.len());

    // Load cache manifest
    let cache_dir = Path::new("beans_feature_cache_112d");
    let cache_manifest_path = cache_dir.join("cache_manifest.json");
    let cache_data = fs::read_to_string(&cache_manifest_path)?;
    let cache_manifest: CacheManifest = serde_json::from_str(&cache_data)?;
    println!("  Cached features: {}", cache_manifest.entries.len());

    // Load all features and labels
    println!("\nLoading features from cache...");
    let mut all_features: Vec<Vec<f32>> = Vec::new();
    let mut all_labels: Vec<String> = Vec::new();

    for sample in &manifest.samples {
        let audio_file = &sample.audio_file;
        let label = if sample.labels.output != "None" {
            sample.labels.output.clone()
        } else {
            format!("task_{}", sample.labels.task)
        };

        if let Some(cache_file) = cache_manifest.entries.get(audio_file) {
            let full_path = cache_dir.join(cache_file);
            if full_path.exists() {
                if let Ok(file) = fs::File::open(&full_path) {
                    let reader = BufReader::new(file);
                    if let Ok(features) = bincode::deserialize_from::<_, Vec<f32>>(reader) {
                        if features.len() == FEATURE_DIM {
                            // Extract only physics features (46D)
                            let physics = slice_physics(&features);
                            all_features.push(physics);
                            all_labels.push(label);
                        }
                    }
                }
            }
        }
    }

    println!("  Loaded {} samples", all_features.len());

    if all_features.is_empty() {
        anyhow::bail!("No features loaded!");
    }

    // Build label mapping
    let mut unique_labels: Vec<String> = all_labels.iter().cloned().collect();
    unique_labels.sort();
    unique_labels.dedup();
    let n_classes = unique_labels.len();
    let mut label_to_idx = HashMap::new();
    for (idx, label) in unique_labels.iter().enumerate() {
        label_to_idx.insert(label.clone(), idx as usize);
    }
    println!("  Classes: {}", n_classes);

    // Convert to ndarray
    println!("\nConverting to ndarray format...");
    let n_samples = all_features.len();
    let mut feature_matrix: Array2<f64> = Array2::zeros((n_samples, PHYSICS_DIM));
    let mut label_array: Array1<usize> = Array1::zeros(n_samples);

    for (i, features) in all_features.iter().enumerate() {
        for (j, &v) in features.iter().enumerate() {
            feature_matrix[[i, j]] = v as f64;
        }
        label_array[i] = *label_to_idx.get(&all_labels[i]).unwrap_or(&0);
    }

    // Split into train/test (90/10)
    let n_train = (n_samples as f32 * 0.9) as usize;
    println!("\nSplitting: {} train, {} test", n_train, n_samples - n_train);

    // Shuffle indices
    let mut indices: Vec<usize> = (0..n_samples).collect();
    for i in 0..indices.len() {
        let j = (rand_u32() as usize) % indices.len();
        indices.swap(i, j);
    }

    let train_indices: Vec<usize> = indices[..n_train].to_vec();
    let test_indices: Vec<usize> = indices[n_train..].to_vec();

    // Create train/test splits
    let train_x = feature_matrix.select_axis(ndarray::Axis(0), &train_indices);
    let train_y = label_array.select(&train_indices);
    let test_x = feature_matrix.select_axis(ndarray::Axis(0), &test_indices);
    let test_y = label_array.select(&test_indices);

    // Standardize features
    println!("\nStandardizing features...");
    let scaler = LinearScaler::standard();
    let train_x_scaled = scaler.fit_transform(&train_x)?;
    let test_x_scaled = scaler.transform(&test_x)?;

    // Create dataset
    let train_dataset = Dataset::new(train_x_scaled.to_owned(), train_y.to_owned());

    // Train Random Forest with balanced class weights
    println!("\nTraining Random Forest...");
    println!("  Max depth: 20");
    println!("  Min samples split: 5");
    println!("  Features: {}D physics", PHYSICS_DIM);

    let model = DecisionTree::params()
        .max_depth(Some(20))
        .min_weight_split(5.0)
        .fit(&train_dataset);

    // Evaluate on test set
    println!("\n╔═══════════════════════════════════════════════════════════════════╗");
    println!("║  Evaluation Results                                               ║");
    println!("╚═══════════════════════════════════════════════════════════════════╝");
    println!();

    // Predict on test set
    let predictions = model.predict(&test_x_scaled);
    
    let mut correct = 0usize;
    for (pred, actual) in predictions.iter().zip(test_y.iter()) {
        if pred == actual {
            correct += 1;
        }
    }
    
    let accuracy = correct as f64 / test_indices.len() as f64 * 100.0;
    println!("Test Accuracy: {:.2}%", accuracy);
    println!("Correct: {} / {}", correct, test_indices.len());

    // Per-class accuracy for top classes
    let mut class_correct: HashMap<usize, usize> = HashMap::new();
    let mut class_total: HashMap<usize, usize> = HashMap::new();

    for (pred, actual) in predictions.iter().zip(test_y.iter()) {
        *class_total.entry(*actual).or_insert(0) += 1;
        if pred == actual {
            *class_correct.entry(*actual).or_insert(0) += 1;
        }
    }

    let mut class_stats: Vec<(usize, usize, usize)> = class_total.iter()
        .map(|(&idx, &total)| {
            let correct = *class_correct.get(&idx).unwrap_or(&0);
            (idx, total, correct)
        })
        .collect();
    class_stats.sort_by(|a, b| b.1.cmp(&a.1));

    println!("\nTop 20 Classes:");
    println!("{:<50} {:>8} {:>8} {:>8}", "Class", "Total", "Correct", "Accuracy");
    println!("{}", "-".repeat(76));

    let idx_to_label: HashMap<usize, &String> = unique_labels.iter()
        .enumerate()
        .map(|(idx, label)| (idx, label))
        .collect();

    for (idx, total, correct) in class_stats.iter().take(20) {
        let label_name = match idx_to_label.get(idx) {
            Some(name) => name.as_str(),
            None => "<unknown>",
        };
        let acc = if *total > 0 {
            (*correct as f64 / *total as f64) * 100.0
        } else {
            0.0
        };
        println!("{:<50} {:>8} {:>8} {:>7.1}%", label_name, total, correct, acc);
    }

    // Save model
    let model_path = "physics_rf_model.json";
    println!("\nSaving model to: {}", model_path);
    
    let model_data = serde_json::json!({
        "model_type": "DecisionTree",
        "feature_dim": PHYSICS_DIM,
        "n_classes": n_classes,
        "labels": unique_labels,
        "accuracy": accuracy,
    });
    fs::write(model_path, serde_json::to_string_pretty(&model_data)?)?;

    println!("\n╔═══════════════════════════════════════════════════════════════════╗");
    println!("║  Summary                                                          ║");
    println!("╠═══════════════════════════════════════════════════════════════════╣");
    println!("║  Architecture:       Random Forest (Physics 46D)                 ║");
    println!("║  Test Accuracy:      {:>8.2}%                                   ║", accuracy);
    println!("║  Total Time:         {:>8.1}s                                    ║", start.elapsed().as_secs_f32());
    println!("╚═══════════════════════════════════════════════════════════════════╝");

    println!("\nHybrid Expert Ensemble Summary:");
    println!("  - Texture NN (66D):  59.88%");
    println!("  - Physics RF (46D):  {:.2}%", accuracy);
    println!("  - Combined ensemble: (run eval_hybrid_ensemble)");

    Ok(())
}

fn rand_u32() -> u32 {
    use std::sync::atomic::{AtomicU64, Ordering};
    static STATE: AtomicU64 = AtomicU64::new(0x853c49e6748fea9b);

    let mut s = STATE.load(Ordering::Relaxed);
    s ^= s >> 12;
    s ^= s << 25;
    s ^= s >> 27;
    STATE.store(s, Ordering::Relaxed);
    (s.wrapping_mul(0x2545F4914F6CDD1D) >> 32) as u32
}
